File: orchestrator-tui/ProxyManager.cs
using Spectre.Console;
using System.Threading;
using System.Threading.Tasks;

namespace Orchestrator;

public static class ProxyManager
{
    // ... (Path variables dan GetProjectRoot() tetap sama) ...
    private static readonly string ProjectRoot = GetProjectRoot();
    private static readonly string ProxySyncDir = Path.Combine(ProjectRoot, "proxysync");
    private static readonly string ProxySyncScript = Path.Combine(ProxySyncDir, "main.py");
    private static readonly string ProxySyncReqs = Path.Combine(ProxySyncDir, "requirements.txt");

     private static string GetProjectRoot()
    {
        var currentDir = new DirectoryInfo(AppContext.BaseDirectory);
        while (currentDir != null) {
            var configDir = Path.Combine(currentDir.FullName, "config");
            var gitignore = Path.Combine(currentDir.FullName, ".gitignore");
            if (Directory.Exists(configDir) && File.Exists(gitignore)) { return currentDir.FullName; }
            currentDir = currentDir.Parent;
        }
        return Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", ".."));
    }

    // Fungsi RunIpAuthorizationOnlyAsync tetap sama
    public static async Task<bool> RunIpAuthorizationOnlyAsync(CancellationToken cancellationToken = default)
    {
        AnsiConsole.MarkupLine("[cyan]--- Menjalankan Auto IP Authorization (ProxySync) ---[/]");
        if (!File.Exists(ProxySyncScript)) {
            AnsiConsole.MarkupLine($"[red]   Error: Skrip ProxySync '{ProxySyncScript}' tidak ditemukan.[/]");
            return false;
        }
        AnsiConsole.MarkupLine("[dim]   Memulai proses IP Auth...[/]");
        try {
            // Panggil dengan flag --ip-auth-only, non-interaktif
            await ShellHelper.RunCommandAsync("python", $"\"{ProxySyncScript}\" --ip-auth-only", ProxySyncDir);
            AnsiConsole.MarkupLine("[green]   ✓ Proses IP Auth selesai.[/]");
            return true;
        } catch (OperationCanceledException) {
             AnsiConsole.MarkupLine("[yellow]   Proses IP Auth dibatalkan.[/]");
             return false;
        } catch (Exception ex) {
            AnsiConsole.MarkupLine($"[red]   ✗ Gagal menjalankan IP Auth: {ex.Message.Split('\n').FirstOrDefault()}[/]");
            return false;
        }
    }

    // --- FUNGSI BARU: Hanya jalankan Test & Save ---
    public static async Task<bool> RunProxyTestAndSaveAsync(CancellationToken cancellationToken = default)
    {
        AnsiConsole.MarkupLine("[cyan]--- Menjalankan Auto Proxy Test & Save (ProxySync) ---[/]");

        if (!File.Exists(ProxySyncScript)) {
            AnsiConsole.MarkupLine($"[red]   Error: Skrip ProxySync '{ProxySyncScript}' tidak ditemukan.[/]");
            return false;
        }

        // Asumsi dependensi sudah siap (diinstall oleh DeployProxies atau ada di remote)

        AnsiConsole.MarkupLine("[dim]   Memulai proses Test & Save...[/]");
        try
        {
            // Panggil main.py dengan flag --test-and-save-only
            await ShellHelper.RunCommandAsync("python", $"\"{ProxySyncScript}\" --test-and-save-only", ProxySyncDir); // <-- Flag baru
            AnsiConsole.MarkupLine("[green]   ✓ Proses Test & Save selesai. 'success_proxy.txt' mungkin diperbarui.[/]");
            return true; // Anggap sukses jika command selesai tanpa error
        }
        catch (OperationCanceledException) {
             AnsiConsole.MarkupLine("[yellow]   Proses Test & Save dibatalkan.[/]");
             return false;
        }
        catch (Exception ex) {
            AnsiConsole.MarkupLine($"[red]   ✗ Gagal menjalankan Test & Save: {ex.Message.Split('\n').FirstOrDefault()}[/]");
            return false;
        }
    }
    // --- AKHIR FUNGSI BARU ---


    // Fungsi DeployProxies tetap sama (untuk menu interaktif)
    public static async Task DeployProxies(CancellationToken cancellationToken = default)
    {
        AnsiConsole.MarkupLine("[bold cyan]--- Menjalankan ProxySync (Lokal - Menu Lengkap) ---[/]");
        // ... (Implementasi DeployProxies tidak berubah) ...
         if (!File.Exists(ProxySyncScript)) {
            AnsiConsole.MarkupLine($"[red]Error: '{ProxySyncScript}' tidak ditemukan.[/]");
            return;
        }
        AnsiConsole.MarkupLine("\n[cyan]1. Menginstal/Update dependensi ProxySync (pip)...[/]");
        try {
            await ShellHelper.RunCommandAsync("pip", $"install --no-cache-dir --upgrade -r \"{ProxySyncReqs}\"", ProxySyncDir);
            AnsiConsole.MarkupLine("[green]   ✓ Dependensi ProxySync siap.[/]");
        } catch (Exception ex) {
            AnsiConsole.MarkupLine($"[red]   Gagal menginstal dependensi: {ex.Message}[/]"); return;
        }
        AnsiConsole.MarkupLine("\n[cyan]2. Menjalankan Menu Interaktif ProxySync...[/]");
        AnsiConsole.MarkupLine("[dim]   (Anda akan masuk ke UI interaktif ProxySync)[/]");
        try {
            await ShellHelper.RunInteractive("python", $"\"{ProxySyncScript}\"", ProxySyncDir, null, cancellationToken);
        } catch (OperationCanceledException) {
             AnsiConsole.MarkupLine("[yellow]   ProxySync dibatalkan oleh user.[/]");
        } catch (Exception ex) {
            AnsiConsole.MarkupLine($"[red]   Gagal menjalankan ProxySync: {ex.Message}[/]");
        }
        AnsiConsole.MarkupLine("\n[bold green]✅ Proses ProxySync selesai.[/]");
        AnsiConsole.MarkupLine("[dim]   File 'proxysync/success_proxy.txt' dan 'config/apilist.txt' mungkin telah diperbarui.[/]");
    }
}

File: py-orchestrator/proxymanager.py
import asyncio
from pathlib import Path
import shutil
import sys
from rich.console import Console
import shell

console = Console()

PROXY_SYNC_PATH = Path("../proxysync").resolve()
CONFIG_SOURCE_PATH = Path("../config").resolve()

async def deploy_proxies():
    console.print("[bold cyan]--- Memulai Proses Deploy Proxy ---[/]")
    
    if not PROXY_SYNC_PATH.is_dir():
        console.print(f"[red]Error: Folder '{PROXY_SYNC_PATH}' tidak ditemukan.[/]")
        console.print("[yellow]Harap jalankan 'Update Semua Bot & Tools' (Menu 2) terlebih dahulu.[/]")
        return
        
    console.print("1. Menyalin file config (apilist.txt, paths.txt)...")
    try:
        shutil.copy(CONFIG_SOURCE_PATH / "apilist.txt", PROXY_SYNC_PATH / "apilist.txt")
        shutil.copy(CONFIG_SOURCE_PATH / "paths.txt", PROXY_SYNC_PATH / "paths.txt")
        console.print("[green]   Salin config berhasil.[/]")
    except Exception as e:
        console.print(f"[red]   Gagal menyalin config: {e}[/]")
        return
        
    console.print("\n2. Menjalankan 'pip install -r requirements.txt' untuk ProxySync...")
    await shell.run_stream_async(
        "pip", "install -r requirements.txt", PROXY_SYNC_PATH
    )
    
    auto_deploy_script = PROXY_SYNC_PATH / "auto_deploy.py"
    
    if auto_deploy_script.exists():
        console.print("\n3. Menjalankan ProxySync (Auto Mode)...")
        console.print("[dim]Download → Convert → Test → Distribute[/]")
        await shell.run_stream_async(
            "python", "auto_deploy.py", PROXY_SYNC_PATH
        )
    else:
        console.print("\n3. Menjalankan ProxySync (Manual Mode)...")
        console.print("[yellow]auto_deploy.py tidak ditemukan, membuka terminal interaktif...[/]")
        shell.run_in_new_terminal(
            "python", "main.py", PROXY_SYNC_PATH
        )
        console.print("[green]Terminal eksternal dibuka.[/]")
        console.print("[dim]Tekan Enter setelah selesai...[/]")
        await asyncio.to_thread(sys.stdin.readline)
        
    console.print("\n[bold green]✅ Proses deploy proxy selesai.[/]")
